src/core/tsi/alts/handshaker/alts_handshaker_client.cc

```diff
@@ -94,45 +117,127 @@ static bool is_handshake_finished_properly(grpc_gcp_HandshakerResp* resp) {
   return false;
 }
 
+static void alts_grpc_handshaker_client_unref(
+    alts_grpc_handshaker_client* client) {
+  if (gpr_unref(&client->refs)) {
+    if (client->base.vtable != nullptr &&
+        client->base.vtable->destruct != nullptr) {
+      client->base.vtable->destruct(&client->base);
+    }
+    grpc_byte_buffer_destroy(client->send_buffer);
+    grpc_byte_buffer_destroy(client->recv_buffer);
+    client->send_buffer = nullptr;
+    client->recv_buffer = nullptr;
+    grpc_metadata_array_destroy(&client->recv_initial_metadata);
+    grpc_slice_unref_internal(client->recv_bytes);
+    grpc_slice_unref_internal(client->target_name);
+    grpc_alts_credentials_options_destroy(client->options);
+    gpr_free(client->buffer);
+    grpc_slice_unref_internal(client->handshake_status_details);
+    gpr_mu_destroy(&client->mu);
+    gpr_free(client);
+  }
+}
+
+static void maybe_complete_tsi_next(
+    alts_grpc_handshaker_client* client, bool receive_status_finished,
+    recv_message_result* pending_recv_message_result) {
+  recv_message_result* r;
+  {
+    grpc_core::MutexLock lock(&client->mu);
+    client->receive_status_finished |= receive_status_finished;
+    if (pending_recv_message_result != nullptr) {
+      GPR_ASSERT(client->pending_recv_message_result == nullptr);
+      client->pending_recv_message_result = pending_recv_message_result;
+    }
+    if (client->pending_recv_message_result == nullptr) {
```

Comment:
Does this branch handle the situation when ALTS handshaker server was unable to handle a client's requests and notify the handshaker client about it?

Reply:
Yes, if I understand the question: if the ALTS handshake server was unable to handle the client's request <i>and</i> the ALTS handshake server was unable to notify the client about it, then what happens is the following:

1) pending RECV_MESSAGE op and pending RECV_STATUS_ON_CLIENT op will both hang
2) After subchannel connection timeout elapses, the subchannel's associated handshake manager will shut down the ongoing security handshake, which will invoke `grpc_call_cancel_internal` on the ALTS RPC
3) pending RECV_MESSAGE and RECV_STATUS_ON_CLIENT ops will then both complete
4) `handle_response_done` invoke this with a non-null `pending_recv_message_result`, having a `status != TSI_OK`
5) `on_status_received` will invoke with this with `receive_status_finished`
6) We'll then invoke the TSI next callback, and the handshake will be terminated.

Note that the added test in `alts_concurrent_connectivity_test.cc`: `TestHandshakeFailsFastWhenHandshakeServerHangsAfterAccepting`, exercises this case.

Reply:
Thanks much for the explanation. It makes sense to me. Could you please also leave this conversation open for future reference?

Reply:
np, will leave this open